### DISCUSSION TITLE
"'build.plugins.plugin.version' for org.codehaus.mojo:build-helper-maven-plugin is missing" in case of multiple source directories

### DISCUSSION THREAD

**Discussion by dhendriks** [eclipse-tycho/tycho#2717] [Q&A] [answered]
If I have multiple source directories in my plugin, I get the following warning:

```
$ mvn clean verify -Dtycho-version=4.0.1
Picked up _JAVA_OPTIONS: -Djava.net.preferIPv4Stack=true
[INFO] Scanning for projects...
[WARNING]
[WARNING] Some problems were encountered while building the effective model for tycho-its-project.compliler:tycho-its-project.compliler.multipleSourceDirs.bundle:eclipse-plugin:1.0.0-SNAPSHOT
[WARNING] 'build.plugins.plugin.version' for org.codehaus.mojo:build-helper-maven-plugin is missing.
[WARNING]
[WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
[WARNING]
[WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
[WARNING]
[INFO] Tycho Version:  4.0.1 (26e55402ca6295ab784674c5053c364ae7b428e7)
[INFO] Tycho Mode:     extension
[INFO] Tycho Builder:  maven
...
```

That is, I have a `src` and `src-gen` folder in my bundle. These folders are configured in `.classpath` and `build.properties`. The bundle has no `pom.xml`, as I use pomless build.

**Comment by laeubi** [comment, marked as answer]
You can add this to your main `pom.xml`

```
<plugin>
	<groupId>org.codehaus.mojo</groupId>
	<artifactId>build-helper-maven-plugin</artifactId>
	<version>3.4.0</version>
</plugin>
```

**Reply by dhendriks** [reply]
I don't add this dependency. It seems Tycho adds it. Couldn't Tycho set the version of that plugin.

**Reply by laeubi** [reply]
> Couldn't Tycho set the version of that plugin.

No as Tycho can't know what version should be used in your build.

**Reply by dhendriks** [reply]
We have this problem in the Eclipse ESCET project. There Tycho generates the `.polyglot.META-INF` file in the directory of my bundle (not sure why it has `META-INF` in the name, `.polyglot.pom.tycho` would make more sense), with the following content:

```
<?xml version="1.0" encoding="UTF-8"?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>
  <parent>
    <groupId>org.eclipse.escet</groupId>
    <artifactId>chi</artifactId>
    <version>1.0.0-SNAPSHOT</version>
    <relativePath>..\.polyglot.pom.tycho</relativePath>
  </parent>
  <artifactId>org.eclipse.escet.chi.parser</artifactId>
  <version>1.0.0-SNAPSHOT</version>
  <packaging>eclipse-plugin</packaging>
  <name>[bundle] ESCET Chi Parser (Incubation)</name>
  <organization>
    <name>Eclipse ESCET</name>
  </organization>
  <build>
    <sourceDirectory>src</sourceDirectory>
    <plugins>
      <plugin>
        <groupId>org.codehaus.mojo</groupId>
        <artifactId>build-helper-maven-plugin</artifactId>
        <executions>
          <execution>
            <id>eclipse-classpath-add-source-1</id>
            <phase>initialize</phase>
            <goals>
              <goal>add-source</goal>
            </goals>
            <configuration>
              <sources>
                <source>src-gen</source>
              </sources>
            </configuration>
          </execution>
        </executions>
      </plugin>
    </plugins>
  </build>
</project>
```

Since this is local for the bundle, can't Tycho generate in this file the version of this plugin, as it will only apply here locally?

**Reply by dhendriks** [reply]
> You can add this to your main `pom.xml` [...]

That seems to work. It still feels like a workaround to me.

**Reply by laeubi** [reply]
> Since this is local for the bundle, can't Tycho generate in this file the version of this plugin, as it will only apply here locally?

But *what* version? If we force a version it will be fixed there and you can never override this (even when a newer/different version is defined in your pom). Therefore the "maven" way is to define this in your root pom.xml in the dependency management section.

**Reply by laeubi** [reply]
This is not a workaround but how maven "works" see:

https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#Dependency_Management

so child modules should usually never define a version but you mange them in the parent.

**Reply by dhendriks** [reply]
Still, I'm using Tycho. Tycho uses something. And then I have to decide on versions of dependencies that I don't directly need, but Tycho needs. This seems to leak implementation details of Tycho to my build configuration.

If I compare this to Eclipse plugins, then if bundle A depends on bundle B, and bundle B depends on bundle C, then B can re-export C, such that A doesn't have to directly depend on C. Much better.

Maybe Maven is just broken, and this is unavoidable, I don't know, but it is not very nice.

**Reply by laeubi** [reply]
Tycho don't "use" something, the tycho-build extension just generates a pom fragment for you but you still need to define at least the `tycho-maven-plugin` version in your root pom and probably more configuration (in this case `build-helper-maven-plugin`), this all has nothing to do with OSGi dependency (and reexport is discouraged anyways and bad practice).

**Comment by dhendriks** [comment]
I've created a reproducer at #2718.
